### PR TITLE
Issue 5295: Bob.jar uses ETag when resolving library dependencies

### LINKED ISSUE #5295: Make editor/bob use correct ETag for downloaded libraries
At least bob is not using the ETag that github provides for the downloaded files.
This means we redownload them for each "resolve" command.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 1]
```diff
@@ -665,6 +665,15 @@ public static void main(String[] args) throws IOException, CompileExceptionError
         System.exit(ret ? 0 : 1);
     }
 
+    public static void main(String[] args) throws IOException, CompileExceptionError, URISyntaxException, LibraryException {
+        try {
+            mainInternal(args);
+        } catch (LibraryException e) {
+            System.err.println(e.getMessage());
+            System.exit(1);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:673]: I _really_ dislike the callstack errors when a failure happens.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 2]
```diff
@@ -1164,39 +1160,43 @@ public void resolveLibUrls(IProgress progress) throws IOException, LibraryExcept
...
-                            if (sha1 != null) {
-                                String serverETag = connection.getHeaderField("ETag");
-                                if (serverETag != null) {
-                                    if (sha1.equals(serverETag.replace("\"", ""))) {
-                                        // Reusing cached library
-                                        serverSha1Match = true;
-                                    }
-                                }
-                            }
+
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:1168]: Not sure we always get a 200 here, but at least we don't get the 400's now.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 3]
```diff
@@ -1164,39 +1160,43 @@ public void resolveLibUrls(IProgress progress) throws IOException, LibraryExcept
...
-
-                            try {
-                                ZipFile zip = new ZipFile(f);
-                                zip.close();
-                            } catch (ZipException e) {
-                                f.delete();
-                                throw new LibraryException(String.format("The file obtained from %s is not a valid zip file", url.toString()), e);
-                            }
-                            logInfo("%2d: Status 200: Stored", i);
+                        if (etag != null && !etag.equals(serverETag)) {
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:1178]: If the previous etag differs, remove the old file on disc.

**Review by britzl — APPROVED** [review]
Great changes!

**Comment by aglitchman** [comment]
I don't know if it is a bug or not.

We made a custom proxy to use our private repositories at _GitLab_ as Defold dependencies. The proxy is very simple, and it doesn't return "ETag" header. This morning (after the release of Defold 1.2.175) our CI builds failed to complete because `bob.jar` refused to download dependencies with the following error:

```
Downloading 4 archive(s)
 0: Downloading https://github.com/subsoap/defos/archive/v2.4.1.zip
 0: Status 304: Already cached
 2147483647% 1: Downloading https://github.com/Insality/druid/archive/0.5.0.zip
 1: Status 304: Already cached
 2: Downloading https://github.com/britzl/ludobits/archive/01c26073e0f7563b99b509e6850b664adeb843d3.zip
 2: Status 304: Already cached
 3: Downloading https://proxy-address-at-netlify?id=16295791
Connection refused by the server at https://proxy-address-at-netlify?id=16295791
```

The addition of the "ETag" header solved the issue.

**Pull request merged**